{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b19ce44f-c3a0-45ad-ae2e-ddcfe01af1d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "78472e76-9a3a-471b-a372-b113eff0a7e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yang.an/mamba/envs/tf2/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n",
      "/Users/yang.an/mamba/envs/tf2/lib/python3.11/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Conditional Formatting extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bait_organism</th>\n",
       "      <th>bait_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_description</th>\n",
       "      <th>majority_protein_acs</th>\n",
       "      <th>fold_change_log2</th>\n",
       "      <th>p_value</th>\n",
       "      <th>fold change_threshold</th>\n",
       "      <th>comparison_batch</th>\n",
       "      <th>fold_change_log2_batch</th>\n",
       "      <th>...</th>\n",
       "      <th>oeproteome_p_value</th>\n",
       "      <th>oeproteome_is_hit</th>\n",
       "      <th>rnaseq_comparison</th>\n",
       "      <th>rnaseq_is_hit</th>\n",
       "      <th>proteome_comparison</th>\n",
       "      <th>proteome_is_hit</th>\n",
       "      <th>phospho_comparison</th>\n",
       "      <th>phospho_is_hit</th>\n",
       "      <th>ubi_comparison</th>\n",
       "      <th>ubi_is_hit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>E</td>\n",
       "      <td>UFD1</td>\n",
       "      <td>Ubiquitin recognition factor in ER-associated ...</td>\n",
       "      <td>Q92890;Q92890-1;Q92890-3</td>\n",
       "      <td>9.969620</td>\n",
       "      <td>6.016913e-276</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_E_vs_B5_others</td>\n",
       "      <td>10.077186</td>\n",
       "      <td>...</td>\n",
       "      <td>0.445602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S247@UFD1[24h:shared(-) 36h:shared(-)] S299@UF...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>E</td>\n",
       "      <td>NPLOC4</td>\n",
       "      <td>Nuclear protein localization protein 4 homolog...</td>\n",
       "      <td>Q8TAT6;Q8TAT6-2</td>\n",
       "      <td>8.854112</td>\n",
       "      <td>3.193217e-20</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_E_vs_B5_others</td>\n",
       "      <td>9.110157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>E</td>\n",
       "      <td>SARS_CoV_E</td>\n",
       "      <td>NO DESCRIPTION</td>\n",
       "      <td>P59637</td>\n",
       "      <td>9.536133</td>\n",
       "      <td>1.491272e-68</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_E_vs_B5_others</td>\n",
       "      <td>9.622996</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K63@SARS_CoV_E[12h:SARS_CoV(+) 24h:SARS_CoV(+)]</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>E</td>\n",
       "      <td>RAB34</td>\n",
       "      <td>Ras-related protein Rab-34, isoform NARR</td>\n",
       "      <td>P0DI83</td>\n",
       "      <td>7.876418</td>\n",
       "      <td>4.422959e-11</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_E_vs_B5_others</td>\n",
       "      <td>8.156219</td>\n",
       "      <td>...</td>\n",
       "      <td>0.052086</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S123@RAB34[36h:shared(-)]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>E</td>\n",
       "      <td>FAM241B</td>\n",
       "      <td>Protein FAM241B...</td>\n",
       "      <td>Q96D05;Q96D05-2</td>\n",
       "      <td>6.863366</td>\n",
       "      <td>3.467090e-04</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_E_vs_B5_others</td>\n",
       "      <td>6.953465</td>\n",
       "      <td>...</td>\n",
       "      <td>0.096263</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2101</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>S</td>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>Palmitoyltransferase ZDHHC5...</td>\n",
       "      <td>Q9C0B5;Q9C0B5-2</td>\n",
       "      <td>7.105674</td>\n",
       "      <td>3.793086e-04</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_S_vs_B4_others</td>\n",
       "      <td>7.465508</td>\n",
       "      <td>...</td>\n",
       "      <td>0.760743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T348@ZDHHC5[12h:shared(+) 24h:SARS_CoV2(+) 36h...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2102</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>S</td>\n",
       "      <td>SARS_CoV_S</td>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>P59594</td>\n",
       "      <td>16.554913</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV_S_vs_B4_others</td>\n",
       "      <td>16.492960</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T215@SARS_CoV_S[36h:SARS_CoV(+)]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>K188@SARS_CoV_S[12h:SARS_CoV(+) 24h:SARS_CoV(+...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2103</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>Palmitoyltransferase ZDHHC5...</td>\n",
       "      <td>Q9C0B5;Q9C0B5-2</td>\n",
       "      <td>7.446649</td>\n",
       "      <td>7.353729e-06</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV2_S_vs_B5_others</td>\n",
       "      <td>7.662380</td>\n",
       "      <td>...</td>\n",
       "      <td>0.565619</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T348@ZDHHC5[12h:shared(+) 24h:SARS_CoV2(+) 36h...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2104</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>GOLGA7</td>\n",
       "      <td>Golgin subfamily A member 7...</td>\n",
       "      <td>Q7Z5G4;Q7Z5G4-3</td>\n",
       "      <td>7.566474</td>\n",
       "      <td>1.689030e-04</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV2_S_vs_B5_others</td>\n",
       "      <td>7.732983</td>\n",
       "      <td>...</td>\n",
       "      <td>0.218128</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K95@GOLGA7[24h:shared(+)]</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2105</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>S</td>\n",
       "      <td>SARS_CoV2_S</td>\n",
       "      <td>Spike glycoprotein</td>\n",
       "      <td>P0DTC2</td>\n",
       "      <td>17.171125</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2</td>\n",
       "      <td>SARS_CoV2_S_vs_B5_others</td>\n",
       "      <td>17.288218</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T791@SARS_CoV2_S[6h:SARS_CoV2(+) 12h:SARS_CoV2...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>K41@SARS_CoV2_S[12h:SARS_CoV2(+) 24h:SARS_CoV2...</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2106 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     bait_organism bait_name    gene_name  \\\n",
       "0         SARS-CoV         E         UFD1   \n",
       "1         SARS-CoV         E       NPLOC4   \n",
       "2         SARS-CoV         E   SARS_CoV_E   \n",
       "3         SARS-CoV         E        RAB34   \n",
       "4         SARS-CoV         E      FAM241B   \n",
       "...            ...       ...          ...   \n",
       "2101      SARS-CoV         S       ZDHHC5   \n",
       "2102      SARS-CoV         S   SARS_CoV_S   \n",
       "2103    SARS-CoV-2         S       ZDHHC5   \n",
       "2104    SARS-CoV-2         S       GOLGA7   \n",
       "2105    SARS-CoV-2         S  SARS_CoV2_S   \n",
       "\n",
       "                                    protein_description  \\\n",
       "0     Ubiquitin recognition factor in ER-associated ...   \n",
       "1     Nuclear protein localization protein 4 homolog...   \n",
       "2                                        NO DESCRIPTION   \n",
       "3              Ras-related protein Rab-34, isoform NARR   \n",
       "4                                    Protein FAM241B...   \n",
       "...                                                 ...   \n",
       "2101                     Palmitoyltransferase ZDHHC5...   \n",
       "2102                                 Spike glycoprotein   \n",
       "2103                     Palmitoyltransferase ZDHHC5...   \n",
       "2104                     Golgin subfamily A member 7...   \n",
       "2105                                 Spike glycoprotein   \n",
       "\n",
       "          majority_protein_acs  fold_change_log2        p_value  \\\n",
       "0     Q92890;Q92890-1;Q92890-3          9.969620  6.016913e-276   \n",
       "1              Q8TAT6;Q8TAT6-2          8.854112   3.193217e-20   \n",
       "2                       P59637          9.536133   1.491272e-68   \n",
       "3                       P0DI83          7.876418   4.422959e-11   \n",
       "4              Q96D05;Q96D05-2          6.863366   3.467090e-04   \n",
       "...                        ...               ...            ...   \n",
       "2101           Q9C0B5;Q9C0B5-2          7.105674   3.793086e-04   \n",
       "2102                    P59594         16.554913   0.000000e+00   \n",
       "2103           Q9C0B5;Q9C0B5-2          7.446649   7.353729e-06   \n",
       "2104           Q7Z5G4;Q7Z5G4-3          7.566474   1.689030e-04   \n",
       "2105                    P0DTC2         17.171125   0.000000e+00   \n",
       "\n",
       "      fold change_threshold          comparison_batch  fold_change_log2_batch  \\\n",
       "0                         2   SARS_CoV_E_vs_B5_others               10.077186   \n",
       "1                         2   SARS_CoV_E_vs_B5_others                9.110157   \n",
       "2                         2   SARS_CoV_E_vs_B5_others                9.622996   \n",
       "3                         2   SARS_CoV_E_vs_B5_others                8.156219   \n",
       "4                         2   SARS_CoV_E_vs_B5_others                6.953465   \n",
       "...                     ...                       ...                     ...   \n",
       "2101                      2   SARS_CoV_S_vs_B4_others                7.465508   \n",
       "2102                      2   SARS_CoV_S_vs_B4_others               16.492960   \n",
       "2103                      2  SARS_CoV2_S_vs_B5_others                7.662380   \n",
       "2104                      2  SARS_CoV2_S_vs_B5_others                7.732983   \n",
       "2105                      2  SARS_CoV2_S_vs_B5_others               17.288218   \n",
       "\n",
       "      ...  oeproteome_p_value oeproteome_is_hit  rnaseq_comparison  \\\n",
       "0     ...            0.445602               0.0                NaN   \n",
       "1     ...            0.009037               0.0                NaN   \n",
       "2     ...                 NaN               NaN                NaN   \n",
       "3     ...            0.052086               0.0                NaN   \n",
       "4     ...            0.096263               0.0                NaN   \n",
       "...   ...                 ...               ...                ...   \n",
       "2101  ...            0.760743               0.0                NaN   \n",
       "2102  ...                 NaN               NaN                NaN   \n",
       "2103  ...            0.565619               0.0                NaN   \n",
       "2104  ...            0.218128               0.0                NaN   \n",
       "2105  ...                 NaN               NaN                NaN   \n",
       "\n",
       "      rnaseq_is_hit  proteome_comparison  proteome_is_hit  \\\n",
       "0               0.0                  NaN              0.0   \n",
       "1               0.0                  NaN              0.0   \n",
       "2               NaN                  NaN              NaN   \n",
       "3               0.0                  NaN              0.0   \n",
       "4               0.0                  NaN              0.0   \n",
       "...             ...                  ...              ...   \n",
       "2101            0.0                  NaN              0.0   \n",
       "2102            NaN                  NaN              0.0   \n",
       "2103            0.0                  NaN              0.0   \n",
       "2104            0.0                  NaN              0.0   \n",
       "2105            NaN                  NaN              0.0   \n",
       "\n",
       "                                     phospho_comparison  phospho_is_hit  \\\n",
       "0     S247@UFD1[24h:shared(-) 36h:shared(-)] S299@UF...             1.0   \n",
       "1                                                   NaN             NaN   \n",
       "2                                                   NaN             NaN   \n",
       "3                             S123@RAB34[36h:shared(-)]             1.0   \n",
       "4                                                   NaN             0.0   \n",
       "...                                                 ...             ...   \n",
       "2101  T348@ZDHHC5[12h:shared(+) 24h:SARS_CoV2(+) 36h...             1.0   \n",
       "2102                   T215@SARS_CoV_S[36h:SARS_CoV(+)]             1.0   \n",
       "2103  T348@ZDHHC5[12h:shared(+) 24h:SARS_CoV2(+) 36h...             1.0   \n",
       "2104                                                NaN             NaN   \n",
       "2105  T791@SARS_CoV2_S[6h:SARS_CoV2(+) 12h:SARS_CoV2...             1.0   \n",
       "\n",
       "                                         ubi_comparison  ubi_is_hit  \n",
       "0                                                   NaN         0.0  \n",
       "1                                                   NaN         0.0  \n",
       "2       K63@SARS_CoV_E[12h:SARS_CoV(+) 24h:SARS_CoV(+)]         1.0  \n",
       "3                                                   NaN         NaN  \n",
       "4                                                   NaN         NaN  \n",
       "...                                                 ...         ...  \n",
       "2101                                                NaN         0.0  \n",
       "2102  K188@SARS_CoV_S[12h:SARS_CoV(+) 24h:SARS_CoV(+...         1.0  \n",
       "2103                                                NaN         0.0  \n",
       "2104                          K95@GOLGA7[24h:shared(+)]         1.0  \n",
       "2105  K41@SARS_CoV2_S[12h:SARS_CoV2(+) 24h:SARS_CoV2...         1.0  \n",
       "\n",
       "[2106 rows x 26 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Labels from https://www.nature.com/articles/s41586-021-03493-4 Stukalov et al. 2021\n",
    "# Direct download: https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-021-03493-4/MediaObjects/41586_2021_3493_MOESM6_ESM.xlsx\n",
    "df = pd.read_excel('41586_2021_3493_MOESM6_ESM.xlsx', sheet_name='A - Significant interactions')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d586a62e-5c8d-43a5-83ab-22dd2eced1eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "612"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract SARS-CoV host factors\n",
    "df_sars = df[~df['gene_name'].str.contains('SARS')]\n",
    "df_sars = df_sars[df_sars['bait_organism'] == 'SARS-CoV']['majority_protein_acs'].str.split(';').explode()\n",
    "df_sars[~df_sars.str.contains('-') & ~df_sars.str.contains('SARS')].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cf606157-58c0-4215-88d5-3b26a12ae536",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = pd.read_csv('../benchmark/prediction_values_all_models.csv.zip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f86693b0-2529-4113-b71b-dcb73efca455",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions['SARS-CoV'] = predictions['protein_ac'].isin(df[df['bait_organism'] == 'SARS-CoV']['majority_protein_acs'].str.split(';').explode().tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e6fdd8b0-17f0-4db3-9507-745c47c5159e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus</th>\n",
       "      <th>AUROC</th>\n",
       "      <th>APS</th>\n",
       "      <th>F1-score</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>Precision@50</th>\n",
       "      <th>Precision@100</th>\n",
       "      <th>Precision@200</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Virus  AUROC   APS  F1-score  Precision  Recall  Precision@50  \\\n",
       "0  SARS-CoV    0.8  0.11      0.19       0.12    0.42          0.16   \n",
       "\n",
       "   Precision@100  Precision@200  \n",
       "0           0.19            0.2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ideal threshold from SARS-CoV-2\n",
    "from sklearn.metrics import roc_auc_score, precision_recall_curve, f1_score, accuracy_score, precision_score, average_precision_score, recall_score, precision_recall_curve\n",
    "import numpy as np\n",
    "\n",
    "modality = 'tf_finetuned'\n",
    "config = 'tf_finetuned_ensemble'\n",
    "Ks = [50, 100, 200]\n",
    "results = []\n",
    "\n",
    "cov = 'SARS-CoV'\n",
    "# Best threshold based on SARS-CoV-2\n",
    "val_subset = predictions[predictions[f'group_split_0'].isin(['train', 'val'])].copy()\n",
    "\n",
    "precision, recall, thresholds = precision_recall_curve(val_subset['label'] == '1.0', val_subset[f'{modality}_ensemble'])\n",
    "f1_scores = 2*precision*recall / (precision+recall)\n",
    "best_threshold = thresholds[f1_scores.argmax()]\n",
    "\n",
    "test_subset = predictions[predictions['group_split_0'] == 'test'].copy()\n",
    "test_subset = predictions.copy()\n",
    "\n",
    "result = {}\n",
    "result.update({\n",
    "          'Virus': cov,\n",
    "          f'AUROC': roc_auc_score(test_subset[cov], test_subset[config]),\n",
    "          f'APS': average_precision_score(test_subset[cov], test_subset[config]),\n",
    "          f'F1-score': f1_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "          f'Precision': precision_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "          f'Recall': recall_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "         })\n",
    "\n",
    "for K in Ks:\n",
    "    test_subset_top_k = test_subset.loc[test_subset[config].nlargest(K).index]\n",
    "    result.update({f'Precision@{K}': precision_score(test_subset_top_k[cov], np.ones_like(test_subset_top_k[config]))})\n",
    "\n",
    "\n",
    "results.append(result)\n",
    "\n",
    "results = pd.DataFrame(results)\n",
    "results.round(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6586a52b-c151-493d-bd15-7bcaadc119cf",
   "metadata": {},
   "source": [
    "# HIV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1804473f-2eb1-4b53-b4e2-21b3cbeab90a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th>number of guides</th>\n",
       "      <th>neg|MAGeCK score</th>\n",
       "      <th>neg|p-value</th>\n",
       "      <th>neg|fdr (false discovery rate)</th>\n",
       "      <th>neg|rank</th>\n",
       "      <th>neg|lfc (log fold change)</th>\n",
       "      <th>pos|MAGeCK score</th>\n",
       "      <th>pos|p-value</th>\n",
       "      <th>pos|fdr</th>\n",
       "      <th>pos|rank</th>\n",
       "      <th>pos|lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UBE2M</td>\n",
       "      <td>8</td>\n",
       "      <td>9.850000e-10</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.623100</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>1052</td>\n",
       "      <td>-1.623100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CCNT1</td>\n",
       "      <td>8</td>\n",
       "      <td>1.430000e-09</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>2</td>\n",
       "      <td>-2.613600</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>1054</td>\n",
       "      <td>-2.613600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PPP2CA</td>\n",
       "      <td>8</td>\n",
       "      <td>1.010000e-08</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.859250</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>1053</td>\n",
       "      <td>-0.859250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZBTB7A</td>\n",
       "      <td>8</td>\n",
       "      <td>1.610000e-08</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.503500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>1051</td>\n",
       "      <td>-1.503500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRNAU1AP</td>\n",
       "      <td>8</td>\n",
       "      <td>1.010000e-07</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>5</td>\n",
       "      <td>-0.525600</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>1049</td>\n",
       "      <td>-0.525600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>SHISAL2A</td>\n",
       "      <td>8</td>\n",
       "      <td>3.108300e-01</td>\n",
       "      <td>0.596350</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>482</td>\n",
       "      <td>-0.027965</td>\n",
       "      <td>0.280210</td>\n",
       "      <td>0.570040</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>593</td>\n",
       "      <td>-0.027965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>RNF17</td>\n",
       "      <td>6</td>\n",
       "      <td>3.184000e-01</td>\n",
       "      <td>0.565400</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>486</td>\n",
       "      <td>-0.211300</td>\n",
       "      <td>0.560600</td>\n",
       "      <td>0.774400</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>774</td>\n",
       "      <td>-0.211300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>ELOVL1</td>\n",
       "      <td>8</td>\n",
       "      <td>3.207900e-01</td>\n",
       "      <td>0.605520</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>489</td>\n",
       "      <td>0.069355</td>\n",
       "      <td>0.411670</td>\n",
       "      <td>0.695810</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>678</td>\n",
       "      <td>0.069355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>DYRK1A</td>\n",
       "      <td>8</td>\n",
       "      <td>3.208100e-01</td>\n",
       "      <td>0.605520</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>490</td>\n",
       "      <td>-0.008780</td>\n",
       "      <td>0.005253</td>\n",
       "      <td>0.027557</td>\n",
       "      <td>0.543204</td>\n",
       "      <td>49</td>\n",
       "      <td>-0.008780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>RFTN1</td>\n",
       "      <td>8</td>\n",
       "      <td>3.215900e-01</td>\n",
       "      <td>0.606300</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>491</td>\n",
       "      <td>-0.113030</td>\n",
       "      <td>0.815940</td>\n",
       "      <td>0.932770</td>\n",
       "      <td>0.999995</td>\n",
       "      <td>882</td>\n",
       "      <td>-0.113030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>368 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene id  number of guides  neg|MAGeCK score  neg|p-value  \\\n",
       "0       UBE2M                 8      9.850000e-10     0.000005   \n",
       "1       CCNT1                 8      1.430000e-09     0.000005   \n",
       "2      PPP2CA                 8      1.010000e-08     0.000005   \n",
       "3      ZBTB7A                 8      1.610000e-08     0.000005   \n",
       "4    TRNAU1AP                 8      1.010000e-07     0.000005   \n",
       "..        ...               ...               ...          ...   \n",
       "481  SHISAL2A                 8      3.108300e-01     0.596350   \n",
       "485     RNF17                 6      3.184000e-01     0.565400   \n",
       "488    ELOVL1                 8      3.207900e-01     0.605520   \n",
       "489    DYRK1A                 8      3.208100e-01     0.605520   \n",
       "490     RFTN1                 8      3.215900e-01     0.606300   \n",
       "\n",
       "     neg|fdr (false discovery rate)  neg|rank  neg|lfc (log fold change)  \\\n",
       "0                          0.000381         1                  -1.623100   \n",
       "1                          0.000381         2                  -2.613600   \n",
       "2                          0.000381         3                  -0.859250   \n",
       "3                          0.000381         4                  -1.503500   \n",
       "4                          0.000381         5                  -0.525600   \n",
       "..                              ...       ...                        ...   \n",
       "481                        0.999995       482                  -0.027965   \n",
       "485                        0.999995       486                  -0.211300   \n",
       "488                        0.999995       489                   0.069355   \n",
       "489                        0.999995       490                  -0.008780   \n",
       "490                        0.999995       491                  -0.113030   \n",
       "\n",
       "     pos|MAGeCK score  pos|p-value   pos|fdr  pos|rank   pos|lfc  \n",
       "0            1.000000     1.000000  0.999995      1052 -1.623100  \n",
       "1            1.000000     1.000000  0.999995      1054 -2.613600  \n",
       "2            1.000000     1.000000  0.999995      1053 -0.859250  \n",
       "3            1.000000     1.000000  0.999995      1051 -1.503500  \n",
       "4            1.000000     1.000000  0.999995      1049 -0.525600  \n",
       "..                ...          ...       ...       ...       ...  \n",
       "481          0.280210     0.570040  0.999995       593 -0.027965  \n",
       "485          0.560600     0.774400  0.999995       774 -0.211300  \n",
       "488          0.411670     0.695810  0.999995       678  0.069355  \n",
       "489          0.005253     0.027557  0.543204        49 -0.008780  \n",
       "490          0.815940     0.932770  0.999995       882 -0.113030  \n",
       "\n",
       "[368 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Labels Supp Table 1 from https://journals.asm.org/doi/10.1128/mbio.00009-23 Montoya et al. 2023\n",
    "# Direct download: https://journals.asm.org/doi/suppl/10.1128/mbio.00009-23/suppl_file/mbio.00009-23-st001.xlsx\n",
    "# The file itself is corrupted and cannot be loaded directly in pandas. Please open using Excel and save the first sheet as csv\n",
    "\n",
    "df_hiv = pd.read_csv('mbio.00009-23-st001.csv')\n",
    "df_hiv = df_hiv[~df_hiv['gene id'].str.contains('synNTC')]\n",
    "df_hiv = df_hiv.iloc()[:368]  # Number used in Montoya et al. 2023\n",
    "df_hiv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "78757597-56fd-47cb-92dd-aaccb8e7ed54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus</th>\n",
       "      <th>AUROC</th>\n",
       "      <th>APS</th>\n",
       "      <th>F1-score</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>Precision@50</th>\n",
       "      <th>Precision@100</th>\n",
       "      <th>Precision@200</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIV</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Virus  AUROC   APS  F1-score  Precision  Recall  Precision@50  \\\n",
       "0  SARS-CoV   0.80  0.11      0.19       0.12    0.42          0.16   \n",
       "1       HIV   0.64  0.03      0.05       0.03    0.18          0.06   \n",
       "\n",
       "   Precision@100  Precision@200  \n",
       "0           0.19           0.20  \n",
       "1           0.05           0.03  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ideal threshold from SARS-CoV-2\n",
    "from sklearn.metrics import roc_auc_score, precision_recall_curve, f1_score, accuracy_score, precision_score, average_precision_score, recall_score, precision_recall_curve\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "predictions['HIV'] = predictions['gene'].isin(df_hiv['gene id'])\n",
    "\n",
    "modality = 'tf_finetuned'\n",
    "config = 'tf_finetuned_ensemble'\n",
    "Ks = [50, 100, 200]\n",
    "results = []\n",
    "\n",
    "for cov in ['SARS-CoV', 'HIV']:\n",
    "    # Best threshold based on SARS-CoV-2\n",
    "    val_subset = predictions[predictions[f'group_split_0'].isin(['train', 'val'])].copy()\n",
    "    \n",
    "    precision, recall, thresholds = precision_recall_curve(val_subset['label'] == '1.0', val_subset[f'{modality}_ensemble'])\n",
    "    f1_scores = 2*precision*recall / (precision+recall)\n",
    "    best_threshold = thresholds[f1_scores.argmax()]\n",
    "    \n",
    "    test_subset = predictions[predictions['group_split_0'] == 'test'].copy()\n",
    "    test_subset = predictions.copy()\n",
    "    # test_subset = predictions[predictions['gene'].isin(df_hiv2['gene id'])].copy()\n",
    "    \n",
    "    result = {}\n",
    "    result.update({\n",
    "              'Virus': cov,\n",
    "              f'AUROC': roc_auc_score(test_subset[cov], test_subset[config]),\n",
    "              f'APS': average_precision_score(test_subset[cov], test_subset[config]),\n",
    "              f'F1-score': f1_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "              f'Precision': precision_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "              f'Recall': recall_score(test_subset[cov], test_subset[config] > best_threshold, ),\n",
    "             })\n",
    "\n",
    "    for K in Ks:\n",
    "        test_subset_top_k = test_subset.loc[test_subset[config].nlargest(K).index]\n",
    "        result.update({f'Precision@{K}': precision_score(test_subset_top_k[cov], np.ones_like(test_subset_top_k[config]))})\n",
    "\n",
    "\n",
    "    results.append(result)\n",
    "\n",
    "results = pd.DataFrame(results)\n",
    "results.round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf2e7a93-3e8c-4125-bd00-cfaac95a31e3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mamba-tf2]",
   "language": "python",
   "name": "conda-env-mamba-tf2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
